File: worker/church_config.js
// Public deployment settings, never credentials. The generator validates the
// same schema before embedding it into Flutter, this Worker and Firestore rules.
import defaultConfig from './generated_config.js';
import { HttpError } from './firebase_user.js';

const SERVICE_ID = /^[A-Za-z][A-Za-z0-9_-]{0,63}$/;
const FEATURES = ['calendar', 'photoImport', 'pushNotifications', 'lineNotifications'];
const ICONS = ['favicon', 'icon192', 'icon512', 'maskable192', 'maskable512'];

function object(value, keys, at) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`${at} must be an object`);
  }
  if (Object.keys(value).some((key) => !keys.includes(key))) {
    throw new Error(`${at} contains an unknown field (credentials belong in secrets, not church config)`);
  }
}

function text(value, at, max = 100, empty = false) {
  if (typeof value !== 'string' || (!empty && !value.trim()) || value.length > max || /[\u0000-\u001f\u007f]/u.test(value)) {
    throw new Error(`${at} must be ${empty ? 'a' : 'a nonempty'} string of at most ${max} characters`);
  }
}

function boolean(value, at) {
  if (typeof value !== 'boolean') throw new Error(`${at} must be a boolean`);
}

function httpsUrl(value, at) {
  text(value, at, 2048, true);
  if (!value) return;
  let url;
  try { url = new URL(value); } catch { throw new Error(`${at} must be an HTTPS URL`); }
  if (url.protocol !== 'https:' || url.username || url.password) {
    throw new Error(`${at} must be an HTTPS URL without credentials`);
  }
}

export function validateChurchConfig(config) {
  object(config, ['schemaVersion', 'appName', 'shortName', 'timeZone', 'services', 'features', 'devotional', 'icons'], 'config');
  if (config.schemaVersion !== 1) throw new Error('schemaVersion must be 1');
  text(config.appName, 'appName');
  text(config.shortName, 'shortName', 40);
  text(config.timeZone, 'timeZone', 100);
  try {
    new Intl.DateTimeFormat('en-US', { timeZone: config.timeZone }).format();
  } catch {
    throw new Error('timeZone must be an IANA time zone');
  }
  // Intl also accepts fixed offsets on newer runtimes; the app requires IANA.
  if (/^[+-]/.test(config.timeZone)) throw new Error('timeZone must be an IANA time zone');
  if (!Array.isArray(config.services) || config.services.length < 1 || config.services.length > 20) {
    throw new Error('services must contain 1 to 20 services');
  }
  const ids = new Set();
  for (const service of config.services) {
    object(service, ['id', 'label', 'name', 'weekday', 'enabled'], 'service');
    if (typeof service.id !== 'string' || !SERVICE_ID.test(service.id) || ['__proto__', 'constructor', 'prototype'].includes(service.id)) {
      throw new Error('service.id must be a stable identifier matching [A-Za-z][A-Za-z0-9_-]{0,63}');
    }
    if (ids.has(service.id)) throw new Error(`duplicate service id: ${service.id}`);
    ids.add(service.id);
    text(service.label, 'service.label', 40);
    text(service.name, 'service.name');
    if (!Number.isInteger(service.weekday) || service.weekday < 1 || service.weekday > 7) {
      throw new Error('service.weekday must be an integer from 1 (Monday) to 7 (Sunday)');
    }
    boolean(service.enabled, 'service.enabled');
  }
  object(config.features, FEATURES, 'features');
  for (const key of FEATURES) boolean(config.features[key], `features.${key}`);
  if (config.features.lineNotifications && !config.features.calendar) {
    throw new Error('lineNotifications requires calendar');
  }
  object(config.devotional, ['enabled', 'dataUrl', 'linkUrl', 'sourceName', 'fetchUrl', 'fetchFormat'], 'devotional');
  boolean(config.devotional.enabled, 'devotional.enabled');
  httpsUrl(config.devotional.dataUrl, 'devotional.dataUrl');
  httpsUrl(config.devotional.linkUrl, 'devotional.linkUrl');
  text(config.devotional.sourceName, 'devotional.sourceName');
  httpsUrl(config.devotional.fetchUrl ?? '', 'devotional.fetchUrl');
  if (!['json', 'dailyBibleHtml'].includes(config.devotional.fetchFormat ?? 'json')) {
    throw new Error('devotional.fetchFormat must be json or dailyBibleHtml');
  }
  if (config.devotional.enabled && (!config.devotional.dataUrl || !config.devotional.linkUrl)) {
    throw new Error('devotional.dataUrl and linkUrl are required when enabled');
  }
  object(config.icons, ICONS, 'icons');
  for (const key of ICONS) {
    const path = config.icons[key];
    text(path, `icons.${key}`, 240);
    if (!/^[A-Za-z0-9_-]+(?:\/[A-Za-z0-9_-]+)*\.png$/.test(path)) {
      throw new Error(`icons.${key} must be a safe relative PNG path`);
    }
  }
  return config;
}

// A symbol cannot be supplied by an HTTP request or a Cloudflare JSON binding.
// It lets unit tests exercise a different deployment without editing sources.
export const TEST_CHURCH_CONFIG = Symbol('test church configuration');

function sortedJsonValue(value) {
  if (Array.isArray(value)) return value.map(sortedJsonValue);
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.keys(value).sort().map((key) => [key, sortedJsonValue(value[key])]));
  }
  return value;
}

function canonicalConfiguration(config) {
  return JSON.stringify(sortedJsonValue({
    ...config,
    devotional: { ...config.devotional, fetchUrl: config.devotional.fetchUrl ?? '', fetchFormat: config.devotional.fetchFormat ?? 'json' },
  }));
}

// Runtime bindings may repeat the generated configuration, but must never
// change its timezone, features, services or branding independently of Flutter
// and the rules. Object key order and optional fetch defaults are immaterial.
const generatedConfiguration = canonicalConfiguration(defaultConfig);
let cachedRaw;
let cachedConfig;
export function churchConfig(env) {
  if (env?.[TEST_CHURCH_CONFIG]) return validateChurchConfig(env[TEST_CHURCH_CONFIG]);
  if (env?.CHURCH_CONFIG_JSON === undefined) return defaultConfig;
  const raw = env.CHURCH_CONFIG_JSON;
  if (raw === cachedRaw) return cachedConfig;
  try {
    if (typeof raw !== 'string') throw new Error('CHURCH_CONFIG_JSON must be a JSON string');
    const config = validateChurchConfig(JSON.parse(raw));
    if (canonicalConfiguration(config) !== generatedConfiguration) {
      throw new Error('runtime config differs from generated deployment; prepare and deploy from one configuration source');
    }
    cachedRaw = raw;
    cachedConfig = config;
    return config;
  } catch (error) {
    console.error('invalid church configuration', error.message);
    throw new HttpError(500, '教會設定有誤，請聯絡管理員');
  }
}

export function requireFeature(env, feature) {
  if (churchConfig(env).features[feature] !== true) {
    throw new HttpError(403, '此教會尚未啟用這項功能');
  }
}

export function dateKeyInZone(date, timeZone) {
  const parts = new Intl.DateTimeFormat('en-US', {
    timeZone, year: 'numeric', month: '2-digit', day: '2-digit',
  }).formatToParts(date);
  const part = (name) => parts.find((entry) => entry.type === name).value;
  return `${part('year')}-${part('month')}-${part('day')}`;
}

File: lib/core/config/church_config.dart
import 'dart:convert';

import '../time/time_zone_database.dart';
import 'default_church_config.dart';

/// 部署者提供的公開設定。憑證不屬於這份設定，也不應進入前端 bundle。
class ChurchConfig {
  ChurchConfig._({
    required this.appName,
    required this.shortName,
    required this.timeZone,
    required this.services,
    required this.features,
    required this.devotional,
    required this.icons,
  });

  static ChurchConfig current = ChurchConfig.fromJson(
    jsonDecode(
          const String.fromEnvironment(
            'CHURCH_CONFIG_JSON',
            defaultValue: defaultChurchConfigJson,
          ),
        )
        as Map<String, dynamic>,
  );

  final String appName;
  final String shortName;
  final String timeZone;
  final List<ServiceDefinition> services;
  final ChurchFeatures features;
  final DevotionalConfig devotional;
  final Map<String, String> icons;

  ServiceDefinition? service(String id) {
    for (final service in services) {
      if (service.id == id) return service;
    }
    return null;
  }

  factory ChurchConfig.fromJson(Map<String, dynamic> json) {
    _checkKeys(json, const [
      'schemaVersion',
      'appName',
      'shortName',
      'timeZone',
      'services',
      'features',
      'devotional',
      'icons',
    ]);
    if (json['schemaVersion'] != 1) {
      throw const FormatException('不支援的教會設定版本');
    }
    final timeZone = _text(json, 'timeZone');
    try {
      timeZoneLocation(timeZone);
    } catch (_) {
      throw FormatException('無效的 IANA 時區：$timeZone');
    }
    final rawServices = json['services'];
    if (rawServices is! List ||
        rawServices.isEmpty ||
        rawServices.length > 20) {
      throw const FormatException('services 必須包含 1 至 20 種聚會');
    }
    final services = rawServices
        .map((raw) => ServiceDefinition.fromJson(_object(raw, 'service')))
        .toList();
    if (services.map((s) => s.id).toSet().length != services.length) {
      throw const FormatException('聚會 ID 不可重複');
    }
    final rawIcons = _object(json['icons'], 'icons');
    _checkKeys(rawIcons, const [
      'favicon',
      'icon192',
      'icon512',
      'maskable192',
      'maskable512',
    ]);
    final icons = rawIcons.map((key, value) {
      if (value is! String ||
          value.length > 240 ||
          !RegExp(
            r'^[A-Za-z0-9_-]+(?:/[A-Za-z0-9_-]+)*\.png$',
          ).hasMatch(value)) {
        throw FormatException('圖示必須是安全的相對 PNG 路徑：$key');
      }
      return MapEntry(key, value);
    });
    for (final key in [
      'favicon',
      'icon192',
      'icon512',
      'maskable192',
      'maskable512',
    ]) {
      if (!icons.containsKey(key)) throw FormatException('缺少圖示：$key');
    }
    return ChurchConfig._(
      appName: _text(json, 'appName'),
      shortName: _text(json, 'shortName', max: 40),
      timeZone: timeZone,
      services: List.unmodifiable(services),
      features: ChurchFeatures.fromJson(_object(json['features'], 'features')),
      devotional: DevotionalConfig.fromJson(
        _object(json['devotional'], 'devotional'),
      ),
      icons: Map.unmodifiable(icons),
    );
  }

  Map<String, dynamic> toJson() => {
    'schemaVersion': 1,
    'appName': appName,
    'shortName': shortName,
    'timeZone': timeZone,
    'services': services.map((s) => s.toJson()).toList(),
    'features': features.toJson(),
    'devotional': devotional.toJson(),
    'icons': icons,
  };
}

class ServiceDefinition {
  const ServiceDefinition({
    required this.id,
    required this.label,
    required this.name,
    required this.weekday,
    required this.enabled,
  });

  final String id;
  final String label;
  final String name;
  final int weekday;
  final bool enabled;

  factory ServiceDefinition.fromJson(Map<String, dynamic> json) {
    _checkKeys(json, const ['id', 'label', 'name', 'weekday', 'enabled']);
    final id = _text(json, 'id', max: 64);
    if (!RegExp(r'^[A-Za-z][A-Za-z0-9_-]{0,63}$').hasMatch(id) ||
        const ['constructor', 'prototype'].contains(id)) {
      throw FormatException('無效的聚會 ID：$id');
    }
    final weekday = json['weekday'];
    if (weekday is! int || weekday < 1 || weekday > 7) {
      throw FormatException('聚會 $id 的 weekday 必須是 1 至 7');
    }
    return ServiceDefinition(
      id: id,
      label: _text(json, 'label', max: 40),
      name: _text(json, 'name'),
      weekday: weekday,
      enabled: _flag(json, 'enabled'),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'label': label,
    'name': name,
    'weekday': weekday,
    'enabled': enabled,
  };
}

class ChurchFeatures {
  const ChurchFeatures({
    this.calendar = false,
    this.photoImport = false,
    this.pushNotifications = false,
    this.lineNotifications = false,
  });

  final bool calendar;
  final bool photoImport;
  final bool pushNotifications;
  final bool lineNotifications;

  factory ChurchFeatures.fromJson(Map<String, dynamic> json) {
    _checkKeys(json, const [
      'calendar',
      'photoImport',
      'pushNotifications',
      'lineNotifications',
    ]);
    final result = ChurchFeatures(
      calendar: _flag(json, 'calendar'),
      photoImport: _flag(json, 'photoImport'),
      pushNotifications: _flag(json, 'pushNotifications'),
      lineNotifications: _flag(json, 'lineNotifications'),
    );
    if (result.lineNotifications && !result.calendar) {
      throw const FormatException('LINE 行事曆通知需要啟用行事曆');
    }
    return result;
  }

  Map<String, dynamic> toJson() => {
    'calendar': calendar,
    'photoImport': photoImport,
    'pushNotifications': pushNotifications,
    'lineNotifications': lineNotifications,
  };
}

class DevotionalConfig {
  const DevotionalConfig({
    required this.enabled,
    required this.dataUrl,
    required this.linkUrl,
    required this.sourceName,
    this.fetchUrl = '',
    this.fetchFormat = 'json',
  });

  final bool enabled;
  final String dataUrl;
  final String linkUrl;
  final String sourceName;
  final String fetchUrl;
  final String fetchFormat;

  factory DevotionalConfig.fromJson(Map<String, dynamic> json) {
    _checkKeys(json, const [
      'enabled',
      'dataUrl',
      'linkUrl',
      'sourceName',
      'fetchUrl',
      'fetchFormat',
    ]);
    final enabled = _flag(json, 'enabled');
    final format = json['fetchFormat'] ?? 'json';
    if (format != 'json' && format != 'dailyBibleHtml') {
      throw const FormatException('不支援的每日靈糧抓取格式');
    }
    return DevotionalConfig(
      enabled: enabled,
      dataUrl: _httpsUrl(json, 'dataUrl', isRequired: enabled),
      linkUrl: _httpsUrl(json, 'linkUrl', isRequired: enabled),
      sourceName: _text(json, 'sourceName'),
      fetchUrl: _httpsUrl(
        {'fetchUrl': json['fetchUrl'] ?? ''},
        'fetchUrl',
        isRequired: false,
      ),
      fetchFormat: format as String,
    );
  }

  Map<String, dynamic> toJson() => {
    'enabled': enabled,
    'dataUrl': dataUrl,
    'linkUrl': linkUrl,
    'sourceName': sourceName,
    'fetchUrl': fetchUrl,
    'fetchFormat': fetchFormat,
  };
}

Map<String, dynamic> _object(Object? value, String field) {
  if (value is! Map<String, dynamic>) {
    throw FormatException('$field 必須是物件');
  }
  return value;
}

void _checkKeys(Map<String, dynamic> json, List<String> allowed) {
  if (json.keys.any((key) => !allowed.contains(key))) {
    throw const FormatException('不支援的設定欄位；憑證不能放在公開設定檔');
  }
}

String _text(Map<String, dynamic> json, String field, {int max = 100}) {
  final value = json[field];
  if (value is! String ||
      value.trim().isEmpty ||
      value.length > max ||
      RegExp(r'[\x00-\x1f\x7f]').hasMatch(value)) {
    throw FormatException('$field 必須是 1 至 $max 字的純文字');
  }
  return value;
}

bool _flag(Map<String, dynamic> json, String field) {
  final value = json[field];
  if (value is! bool) throw FormatException('$field 必須是布林值');
  return value;
}

String _httpsUrl(
  Map<String, dynamic> json,
  String field, {
  required bool isRequired,
}) {
  final value = json[field];
  if (value is! String ||
      value.length > 2048 ||
      RegExp(r'[\x00-\x1f\x7f]').hasMatch(value)) {
    throw FormatException('$field 必須是安全的 URL 字串');
  }
  if (value.isEmpty && !isRequired) return '';
  final uri = Uri.tryParse(value);
  if (uri == null ||
      uri.scheme != 'https' ||
      uri.host.isEmpty ||
      uri.userInfo.isNotEmpty) {
    throw FormatException('$field 必須是 HTTPS 網址');
  }
  return value;
}
